### ISSUE TITLE
Delay assumes perfect scheduling

### ISSUE THREAD

**Issue by nitsanw** [engineblock/engineblock#156]
In the rate limiting code [here](https://github.com/engineblock/engineblock/blob/0010c61b76c8db4a11c1749473144fe053fcafc6/eb-api/src/main/java/io/engineblock/activityapi/rates/AverageRateLimiter.java#L155) when c.o. avoiding reporting is enabled we are supposed to (as far as I gather from the code) report the difference between the actual start time and the "ideal" time for this operation. The `delay` should always be `now - scheduledTime` but instead we calculate it to be (comments are mine):
```
        long opScheduleTime = ticksTimeline.getAndAdd(nanos);
        long timelinePosition = lastSeenNanoTime;

// we are behind schedule according to a cached time value, but should report the "real" delay.
        if (opScheduleTime < timelinePosition) {
            isBursting = true;
            return reportCoDelay ? timelinePosition - opScheduleTime : 0L;
        }

        long delay = timelinePosition - opScheduleTime;
// delay at this point is always <= 0, otherwise would have returned above, so this condition seems redundant?
        if (delay <= 0) {
            timelinePosition = getNanoClockTime();
// So, if we hit some large delay we will compute the delay from this last cached value.
            lastSeenNanoTime = timelinePosition;
            delay = timelinePosition - opScheduleTime;

            if (delay < 0) {
                isBursting = false;
// why not compute the positive value upfront and reverse the conditions?
                delay *= -1; 
                try {
// Sleeping accurately for short periods is impossible. 
// See the self calibrating sleeper in bdp: https://github.com/nitsanw/bdp/blob/f152a864b0e70f50f53b58df2d7d629373d542a9/dse-db/src/java/org/apache/cassandra/concurrent/ParkedThreadsMonitor.java#L254
                    Thread.sleep(delay / 1000000, (int) (delay % 1000000L));
                } catch (InterruptedException ignored) {
                }
// this assumes perfect sleeps, which is unrealistic. Consider for instance a multi second GC pause happens during the sleep.
                return 0L;
            }
        }

        return reportCoDelay ? delay : 0L;
```
I think the correct solution is to always return `now - opScheduleTime` when `reportCoDelay == true`

**Comment by jshook** [comment]
This code is become replaced by something that is much simpler and easier to reason about. I'll close this with a pointer to that implementation once it is in a commit.

**Comment by jshook** [comment]
@nitsanw In case you are interested in what happened with this, I converted the rate limiter implementation into a leaky bucket rate limiter. After a moderate effort at trying out various approaches, the method that seemed to be the best balanced was one that delegated to the JVM for lock elision, etc using a simple (but limited) synchronized method on a dedicated filler thread. There is possibility for blocking on exhaustion (wait,notify) but when possible all resource consumers use CAS methods to avoid a worse form of blocking.

It's not perfect, but across the scenarios I was able to test with, it had the least amount of surprises when going from low threads to high and lower contention to higher contention. The previous implementation was too difficult to explain and reason about. But even more troublesome is that when I added the additional features that a tester would want in a rate limiter, it was too inflexible and difficult to test. The current implementation does have issues but I think they can be overcome, and it has the benefit of being explainable even with the advanced features. It works well enough on most hardware, and smarter time-slicing will make it work well in nearly all cases I've seen so far.

If you find another example somewhere that avoids free-wheeling loops and PLL simulation, I'm interested in investigating and learning more about alternatives.

Thanks for your interest in helping on this when you did. Cheers!
